test/test_controllers/test_generic_controller.py

```diff
@@ -0,0 +1,102 @@
+    def __init__(self, components: dict):
+        self.custom_attribute = "abc"
+        super().__init__(components=components)
+
+
+class ExampleControllerWithObservers(GenericController):
+    def __init__(self):
+        super().__init__(components=dict(button=Button(), check_box=Checkbox()))
+
+    def check_box_on_button_click(self, change):
+        if self.check_box.value is True:
+            self.check_box.value = False
+        elif self.check_box.value is False:
+            self.check_box.value = True
+
+    def setup_observers(self):
+        self.button.on_click(self.check_box_on_button_click)
+
+
+class TestGenericController(TestCase):
+    def test_generic_controller_component_controller_assertion(self):
+        components = dict(other_controller=ExampleControllerWithObservers())
+
+        with self.assertRaisesWith(
+            exc_type=ValueError,
+            exc_msg=(
+                "Component 'other_controller' is a type of Controller! "
+                "Please use a MultiController to unpack its components."
+            ),
+        ):
+            ExampleGenericController(components=components)
+
+    def test_generic_controller_name_collision(self):
+        components = dict(button=Button(), check_box=Checkbox(), custom_attribute=123)
+
+        with self.assertRaisesWith(
+            exc_type=KeyError,
+            exc_msg=(
+                "\"This Controller already has an outer attribute with the name 'custom_attribute'! "
+                'Please adjust the reference key to be unique."'
+            ),
+        ):
+            ExampleGenericController(components=components)
+
+    def test_generic_controller_no_components(self):
+        controller = GenericController(components=dict())
+
+        self.assertDictEqual(d1=controller.components, d2=dict())
+        self.assertDictEqual(d1=controller.get_fields(), d2=dict())
+        self.assertTupleEqual(tuple1=controller.children, tuple2=())
+
+    def test_generic_controller_ipywidget_components(self):
+        button = Button()
+        check_box = Checkbox()
+        components = dict(button=button, check_box=check_box)
+        controller = GenericController(components=components)
+
+        expected_children = (button, check_box)
+        self.assertDictEqual(d1=controller.components, d2=components)
+        self.assertDictEqual(d1=controller.get_fields(), d2=components)
+        self.assertTupleEqual(tuple1=controller.children, tuple2=expected_children)
+
+    def test_generic_controller_standard_attributes(self):
+        """Non-widget attributes were not included in the children of the box."""
+        button = Button()
+        check_box = Checkbox()
+        components = dict(
+            button=button,
+            check_box=check_box,
+            some_integer=1,
+            some_string="test",
+            some_float=1.23,
+            some_list=[1, 2, 3],
+            some_dict=dict(a=5, b=6, c=7),
+        )
```

Comment:
I don't really understand why GenericController contains non-widget components

Reply:
Makes it flexible for when you have multiple widget components that can mutually affect other properties - example from the original PR (not updated to this GenericController standard) is the RotationController: https://github.com/catalystneuro/nwbwidgets/blob/add_two_photon_color/nwbwidgets/controllers/image_controllers.py#L7-L29

Two buttons that affect a common value that is then used to determine some aspect of the final visualization.

Also, some attempts to use properties of widgets (an example being the fields of a DropDown) do not update in-place, so having a secondary attribute that is modifiable can be handy as well if you ever want interactions with those properties

Reply:
ok, but why throw all of these into one dictionary? I don't really understand the logic of this PR in several ways so let's talk about it when we meet tomorrow.

Reply:
> ok, but why throw all of these into one dictionary?

For input, they naturally form key/value pairs as attributes. As far as why `components` itself is also saved as an attribute (which might seem redundant), yeah that's largely there for the `MultiController` which has the added behavior of more human-readable nesting of sub-controller components if someone ever wanted an easier direct reference to it as opposed to navigating lists of nested `children`. That is, `some_multi_controller.components["SomeSubController"]` instead of trying to find `SomeSubController` out of potentially several indices of `some_multi_controller.children`, which could get even worse when you start making `MultiControllers` out of `MultiControllers`.

I'd be OK with removing `components` as a stored attribute for the `BaseController` and the `GenericController` and only have it be present for the `MultiController`.


> I don't really understand the logic of this PR in several ways so let's talk about it when we meet tomorrow.

I'd say it's better to ask as many questions as you can in text form so we have a permanent reference to look back on and also bring in @h-mayorquin and @luiztauffer who contributed a lot to early thoughts on designs for this kind of stuff

Reply:
Ah, good point. I can adjust it so that `.components` only stores `ipywidgets` object references. Other attributes can then be retrieved directly or via `get_fields()` (they convenience retrieval for seeing what attributes the class has instead of looking at it's magic `__dict__`)